### PR TITLE
Add Marketplace subscriptions API license error responses

### PR DESCRIPTION
**Pull request by bgrgicak**
### Changes proposed in this Pull Request:

This PR adds error responses to the `marketplace/subscriptions/activate` and `marketplace/subscriptions/deactivate` API endpoints. 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check out this branch.
2. Add `return true` [to this line](https://github.com/woocommerce/woocommerce/blob/7a6ad615ca06e5805def757af004b32f3e230036/plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php#L76) 
3. Request an non-existing license activation
```
curl -X POST --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/activate' \
--form 'product_key="fake"'
```
4. Request an non-existing license deactivation
```
curl -X POST --location 'http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/deactivate' \
--form 'product_key="fake"'
```

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Add Marketplace subscriptions API license error responses.

</details>

### CODE REVIEW

**Comment by andfinally** [comment]
It's strange, I'm getting 404 from the endpoints, even when I make a POST request. For example to `http://localhost:8888/wp-json/wc/v3/marketplace/subscriptions/activate`. My local store is accessible in the browser, and I can see that `register_rest_routes` is being called, so the endpoint should exist.

**Comment by bgrgicak** [comment]
@andfinally can you [update your permalink structure](http://localhost:8888/wp-admin/options-permalink.php) to _Post name_ and try again?

Screenshot 2023-10-12 at 11 20 58

**Comment by andfinally** [comment]
Fantastic, all good now, thanks Bero!

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php [hunk 1]
```diff
@@ -903,7 +903,12 @@ private static function _helper_subscription_activate() {
 			wp_die( 'Could not verify nonce' );
 		}
 
-		$activated = self::activate_helper_subscription( $product_key, $product_id );
+		$activated = false;
```
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: Do we need this line, if the `try ... catch` sets the value of `$activated`? Same with `$deactivated`.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php [hunk 2]
```diff
@@ -928,7 +933,7 @@ private static function _helper_subscription_activate() {
 	public static function activate_helper_subscription( $product_key ) {
 		$subscription = self::get_subscription( $product_key );
 		if ( ! $subscription ) {
-			return false;
+			throw new Exception( __('Subscription not found', 'woocommerce') );
```
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: PhpStorm suggests we add a `@throws Exception` to the function docblock to account for this. (Same with `deactivate_helper_subscription`.) 

It also says that `$activated` is always true when returned on line 993, and suggests we just return `true`.

**Review by andfinally — APPROVED** [review]
Looks good to me and works as expected, thanks! Just a couple of minor queries.

The big question for me is whether this change breaks these operations for older versions of WooCommerce, or when the React marketplace feature is disabled. Unfortunately, none of my subscriptions offer me an activate/deactivate option: they only have a `Download` button. I believe this is because my local site is inaccessible to WCCOM. When we're nearly ready to merge the feature branch, we might be able to test this by building teh plugin and installing it on an external test site.

---------------

LATER

Tried this test on a JN site... the extension activated, but I got a bunch of errors, hopefully unrelated to these changes. 

```
[Thu Oct 12 10:37:19.050132 2023] [proxy:error] [pid 31778:tid 139894257018624] (2)No such file or directory: AH02454: FCGI: attempt to connect to Unix domain socket /srv/users/user4decdd26/run/user4decdd26.php-fpm.sock (localhost:8000) failed
[Thu Oct 12 10:37:19.050202 2023] [proxy_fcgi:error] [pid 31778:tid 139894257018624] [client [ IP number ]:0] AH01079: failed to make connection to backend: httpd-UDS, referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper
[Thu Oct 12 10:37:53.869968 2023] [proxy_fcgi:error] [pid 4195:tid 139894382909184] [client [ IP number ]:0] AH01067: Failed to read FastCGI header - see https://serverpilot.io/community/articles/error-failed-to-read-fastcgi-header.html for more information, referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper&filter=all&wc-helper-status=helper-refreshed
[Thu Oct 12 10:37:53.869996 2023] [proxy_fcgi:error] [pid 4195:tid 139894382909184] (104)Connection reset by peer: [client [ IP number ]:0] AH01075: Error dispatching request to : , referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper&filter=all&wc-helper-status=helper-refreshed
[Thu Oct 12 10:38:10.216717 2023] [proxy:error] [pid 4590:tid 139894488004352] (2)No such file or directory: AH02454: FCGI: attempt to connect to Unix domain socket /srv/users/user4decdd26/run/user4decdd26.php-fpm.sock (localhost:8000) failed
[Thu Oct 12 10:38:10.216757 2023] [proxy_fcgi:error] [pid 4590:tid 139894488004352] [client [ IP number ]:0] AH01079: failed to make connection to backend: httpd-UDS, referer: [ JN address ]/wp-admin/plugins.php?plugin_status=all&paged=1&s
[Thu Oct 12 10:38:19.646815 2023] [proxy_fcgi:error] [pid 4931:tid 139894521575168] [client [ IP number ]:0] AH01067: Failed to read FastCGI header - see https://serverpilot.io/community/articles/error-failed-to-read-fastcgi-header.html for more information, referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper&filter=all&wc-helper-status=helper-refreshed
[Thu Oct 12 10:38:19.646842 2023] [proxy_fcgi:error] [pid 4931:tid 139894521575168] (104)Connection reset by peer: [client [ IP number ]:0] AH01075: Error dispatching request to : , referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper&filter=all&wc-helper-status=helper-refreshed
[Thu Oct 12 10:38:24.426437 2023] [proxy_fcgi:error] [pid 4847:tid 139894513182464] [client [ IP number ]:0] AH01067: Failed to read FastCGI header - see https://serverpilot.io/community/articles/error-failed-to-read-fastcgi-header.html for more information, referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper
[Thu Oct 12 10:38:24.426482 2023] [proxy_fcgi:error] [pid 4847:tid 139894513182464] (104)Connection reset by peer: [client [ IP number ]:0] AH01075: Error dispatching request to : , referer: [ JN address ]/wp-admin/admin.php?page=wc-addons&section=helper
```

**Comment by bgrgicak** [comment]
> Unfortunately, none of my subscriptions offer me an activate/deactivate option: they only have a Download button. I believe this is because my local site is inaccessible to WCCOM. 

@andfinally you should be able to test it locally.
After you download and manually install one of your products, you should see a activate/deactivate toggle.

**Comment by bgrgicak** [comment]
> The big question for me is whether this change breaks these operations for older versions of WooCommerce, or when the React marketplace feature is disabled. 

It should work for both. I made sure to add a try catch to all places that use this code.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php (hunk 2 above)
Comment (bgrgicak) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: [059d447](https://github.com/woocommerce/woocommerce/pull/40703/commits/059d44760a01980d244d1acb0eb679e64c863b42)

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php (hunk 1 above)
Comment (bgrgicak) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: [059d447](https://github.com/woocommerce/woocommerce/pull/40703/commits/059d44760a01980d244d1acb0eb679e64c863b42)

**Review by andfinally — COMMENTED** [review]
I'm now only seeing a few small changes in this diff – whitespace, comment and removing the unsued `Themes` import – so I guess the substantive changes have been made in the feature branch. In this case, it's probably safe to merge this as is, or scrap it.

**Review by Dan-Q — APPROVED** [review]
Behaves as promised!

**Comment by bgrgicak** [comment]
@andfinally somehow this [code is already in the feature branch](https://github.com/woocommerce/woocommerce/pull/40249/files#diff-32b26bf8430b272079f3c981317024f2a7d7848b4198ae7878716cc133d10961R968).
I will merge these linter fixes.

**Pull request merged**